rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,221 @@
+        """List of conflicting actions.
+
+        Returns:
+            List of conflicting actions.
+
+        """
+        return list(self._conflicting_actions.keys())
+
+    @property
+    def has_prior_events(self) -> bool:
+        """Checks if prior events exist.
+
+        Returns:
+            True if anything has happened before this conflict, otherwise False.
+        """
+        return _get_previous_event(self.sliced_states[-1])[0] is not None
+
+    def __str__(self) -> Text:
+        # Describe where the conflict occurs in the stories
+        last_event_type, last_event_name = _get_previous_event(self.sliced_states[-1])
+        if last_event_type:
+            conflict_message = (
+                f"CONFLICT after {last_event_type} '{last_event_name}':\n"
+            )
+        else:
+            conflict_message = f"CONFLICT at the beginning of stories:\n"
+
+        # List which stories are in conflict with one another
+        for action, stories in self._conflicting_actions.items():
+            # Summarize if necessary
+            if len(stories) > 3:
+                # Four or more stories are present
+                conflict_description = (
+                    f"'{stories[0]}' and {len(stories) - 1} other trackers"
+                )
+            else:
+                conflict_description = (
+                    {1: "'{}'", 2: "'{}' and '{}'", 3: "'{}', '{}', and '{}'",}
+                    .get(len(stories))
+                    .format(*stories)
+                )
+
+            conflict_message += f"  {action} predicted in {conflict_description}\n"
+
+        return conflict_message
+
+
+TrackerEventStateTuple = namedtuple(
+    "TrackerEventStateTuple", "tracker event sliced_states"
+)
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> List[StoryConflict]:
+    """Generates a list of StoryConflict objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+    Returns:
+        List of conflicts.
+    """
+
+    # We do this in two steps, to reduce memory consumption:
+
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_dict = _find_conflicting_states(trackers, domain, max_history)
+
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = _build_conflicts_from_states(
+        trackers, domain, max_history, state_action_dict
+    )
+
+    return conflicts
+
+
+def _find_conflicting_states(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> Dict[int, Optional[List[Text]]]:
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_dict = defaultdict(list)
```

Comment:
please avoid hungarian notation
```suggestion
    state_action_mapping = defaultdict(list)
```